### ISSUE TITLE
Implement a maximum pandas compatibility mode

### ISSUE THREAD

**Issue by gshimansky** [modin-project/modin#3411]
Some recent PRs introduced known difference in behavior between Modin and Pandas. I know of at least two:
1. Different order of columns in `groupby.__getitem__` https://github.com/modin-project/modin/pull/3298#discussion_r699226287
2. `Index` instead of `MultiIndex` in `DataFrame.value_counts` https://github.com/modin-project/modin/pull/2730#discussion_r700153986

Workarounds for differences in both cases above are trivial but impair performance. I suggest to add a setting called something like `MODIN_MAXIMUM_PANDAS_COMPAT` that would implement workarounds at the expense of performance and make functions behavior in the closest possible way to how Pandas implements them. In case some user's application breaks when he replaces `import pandas` with `import modin.pandas` such setting would allow to quickly fix problems that happen because of incompatibility.

**Comment by amyskov** [comment]
I think such mode can be very helpful for https://github.com/modin-project/modin/pull/3288. In the case if heterogeneous data is imported, partitions data can have different type and casting it to proper data type can be expensive. In the case if such discrepancy is unimportant, this casting can be skipped.

**Comment by dchigarev** [comment]
Another couple of cases that deviate from pandas:
1. `PandasQueryCompiler.melt` - pandas `melt` function preserves the order of rows of the `id_vars` column, in the current implementation order of rows is chaotic
https://github.com/modin-project/modin/blob/0a642753af2cd2332ac71166bcddc89b0cace9d5/modin/backends/pandas/query_compiler.py#L2027-L2029
2. `PandasQueryCompiler.pivot_table` - the results column labels have to be sorted in ascending order, currently unsorted
https://github.com/modin-project/modin/blob/0a642753af2cd2332ac71166bcddc89b0cace9d5/modin/backends/pandas/query_compiler.py#L2936-L2939

**Comment by dchigarev** [comment]
I think we should start from determining some kind of criterios, that will describe the expected difference between strict and non-strict modes. With their help, we can easily modify our current comparators in tests. I can suggest that the following is not guaranteed in the non-strict mode:
1. Order of rows and columns could be different from pandas. Comparator changes: both axes have to be sorted before comparison with pandas result.
2. Object type of axis labels could be different, however, its content is guarantee to be the same (for except of the order). Comparator changes:  before comparison, cast index object type to align with pandas type if required.
3. Data type of columns could be different. This actually needs to be clarified, we need to agree on which data-types we allow deviations and what kind this deviation could be. I suggest that the initial alloweness table is:

|modin type|pandas type|is allowed|
|-|-|-|
|any numeric|any numeric|+|
|any numeric|object|+|
|object|any numeric|-|
|any datetime|any datetime|+|
|object|any datetime|-|
|any datetime|object|+|

cc @modin-project/modin-core

**Comment by RehanSD** [comment]
Hi @dchigarev @gshimansky @amyskov has this issue been resolved?

**Comment by Garra1980** [comment]
No, this is something we discussed some time ago but never implemented. It is a performance-related feature and we may return to this idea in the future
